diff -ruN linux-4.9.37_original/drivers/clocksource/timer-sp804.c linux-4.9.37_modified/drivers/clocksource/timer-sp804.c
--- linux-4.9.37_original/drivers/clocksource/timer-sp804.c	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.37_modified/drivers/clocksource/timer-sp804.c	2018-07-04 07:50:03.067724305 +0300
@@ -235,6 +235,10 @@
 	writel(0, base + TIMER_CTRL);
 	writel(0, base + TIMER_2_BASE + TIMER_CTRL);
 
+	/* Ensure timer interrupts are clear */
+	writel(1, base + TIMER_INTCLR);
+	writel(1, base + TIMER_2_BASE + TIMER_INTCLR);
+
 	if (initialized || !of_device_is_available(np)) {
 		ret = -EINVAL;
 		goto err;
