diff --git a/drivers/mtd/spi-nor/Kconfig b/drivers/mtd/spi-nor/Kconfig
index 4a682ee..b72cf27 100644
--- a/drivers/mtd/spi-nor/Kconfig
+++ b/drivers/mtd/spi-nor/Kconfig
@@ -59,11 +59,12 @@ config SPI_FSL_QUADSPI
 	  SPI NOR.
 
 config SPI_HISI_SFC
-	tristate "Hisilicon SPI-NOR Flash Controller(SFC)"
-	depends on ARCH_HISI || COMPILE_TEST
+	tristate "Hisilicon FMCV100 SPI-NOR Flash Controller(SFC)"
+	depends on ARCH_HISI || ARCH_HISI_BVT || COMPILE_TEST
 	depends on HAS_IOMEM && HAS_DMA
 	help
-	  This enables support for hisilicon SPI-NOR flash controller.
+	  This enables support for hisilicon flash memory contrller ver100
+	  (FMCV100)- SPI-NOR flash controller.
 
 config SPI_NXP_SPIFI
 	tristate "NXP SPI Flash Interface (SPIFI)"
